framework/src/transfers/MultiAppMeshFunctionTransfer.C

```diff
@@ -74,6 +81,29 @@ MultiAppMeshFunctionTransfer::execute()
 
   getAppInfo();
 
+  // loop over the vector of variables and make the transfer one by one
+  for (unsigned int i = 0; i < _var_size; ++i)
+    transferVariable(i);
+
+  // Make sure all our sends succeeded.
+  for (processor_id_type i_proc = 0; i_proc < n_processors(); ++i_proc)
+  {
+    if (i_proc == processor_id())
+      continue;
+    _send_points[i_proc].wait();
+    _send_evals[i_proc].wait();
+    if (_direction == FROM_MULTIAPP)
+      _send_ids[i_proc].wait();
+  }
```

Comment:
This, now, is kind of weird.  You're postponing wait() until *subsequent* invocations of execute()?

Reply:
No, he's transferring all variables (used to be one, now he's modified this to do multiple), then doing all the waits as part of the original invocation of execute(). Everything should be complete (including waits) before this transfer is complete. 

Are you saying that the waits should be done per variable?

Reply:
@roystgnr If you think the waits should be done per variable, let us know!

Reply:
Well, the important thing is that there should be exactly one wait on a request object per asynchronous operation on that object.
...
And maybe I'm reading this wrong still, but it looks like that's not what's happening?  It looks like you're trying to reuse three vectors of ```n_procs``` requests, for a total of ```3 * n_procs * n_vars``` sends?  In that case, the second variable you transfer will be overwriting the request objects from the first variable, before the wait operates, which means those objects will leak, and those waits won't be completed as expected.

Postponing all the waits until after all the transfers may be a good idea, but if you're doing ```3 * n_procs * n_vars``` asynchronous sends then you need ```3 * n_procs * n_vars``` request objects to keep track of them all.

Reply:
oh yes, I understand your point and I agree with you: as is, ```_send_points```, ```_send_evals``` and ```_send_ids``` will not store the values for each variable. I should make them ```std::vector<std::vector<Parallel::Request>>```

Thank you for pointing that out!